proxy/envoy/config.go

```diff
@@ -119,28 +116,37 @@ func build(instances []*model.ServiceInstance, services []*model.Service,
 	inbound := buildInboundFilters(instances)
 
 	// merge the two sets of route configs
-	configs := make(RouteConfigs)
-	for port, config := range inbound {
-		configs[port] = config
+	routeConfigs := make(RouteConfigs)
+	for port, routeConfig := range inbound {
+		routeConfigs[port] = routeConfig
 	}
 
 	for port, outgoing := range outbound {
-		if incoming, ok := configs[port]; ok {
+		if incoming, ok := routeConfigs[port]; ok {
 			// If the traffic is sent to a service that has instances co-located with the proxy,
 			// we choose the local service instance since we cannot distinguish between inbound and outbound packets.
 			// Note that this may not be a problem if the service port and its endpoint port are distinct.
-			configs[port] = incoming.merge(outgoing)
+			routeConfigs[port] = incoming.merge(outgoing)
 		} else {
-			configs[port] = outgoing
+			routeConfigs[port] = outgoing
 		}
 	}
 
 	// canonicalize listeners and collect clusters
 	clusters := make(Clusters, 0)
 	listeners := make([]*Listener, 0)
-	for port, config := range configs {
-		sort.Sort(HostsByName(config.VirtualHosts))
-		clusters = append(clusters, config.clusters()...)
+	for port, routeConfig := range routeConfigs {
+		sort.Sort(HostsByName(routeConfig.VirtualHosts))
+		clusters = append(clusters, routeConfig.clusters()...)
+
+		filters := buildFaultFilters(config, routeConfig)
+
+		filters = append(filters, Filter{
+			Type:   "decoder",
+			Name:   "router",
+			Config: FilterRouterConfig{},
+		})
```

Comment:
And the only reason I needed to do fault filter generation in the build function was because I needed to get hold of all the clusters per http_conn_filter, so that I can generate the appropriate fault filter blocks. I cannot do that outside the build function (it becomes messy).